### PR TITLE
Gun refactor

### PR DESCRIPTION
**Pull request by metalgearsloth**
How guns work:
- GunComponent exists
- Client attempts to shoot gun n times in a specific tick, raising a predicted event, and raises an event on itself to get ammo
- AmmoProviders provide ammo via subscription and handle their own logic
- Bullets shot. If no bullets shot play empty sound.

AmmoCounter is also ECS driven and on its own component so any gun can have a basic ammo counting by just adding or removing the component; having specific controls is just handled by an event.

I'm putting as much code into shared as I can though because of batteries / actions / do_afters not being predicted some has to be moved (do_afters for reloads etc is tbc).

Ammo is a bit weird as supporting both hitscan and entity shooting required the interface which means launchers will add a dummy AmmoComponent to stuff being launched which immediately gets removed. This also applies to the client spawning its entities then immediately deleting them (at least for now).

ChamberMagazine + Magazine ammo providers just pass most of their events onwards to the magazines so having whatever as a magazine should be fairly simple. The yml for the itemslots is a bit cbt atm so this may change.

Other notes:
- Projectile spawning isn't predicted but the muzzle flash and sound are which I think is the best we can hope for anyway.
- Guns just use whitelists instead of calibers which means you can load many things into guns easily, same with ItemSlots so guns benefit from any bugfixes to those.
- Significantly cut down the number of rsis and sprites, to the point it will probably have a noticeable impact on game startup
- The diff is huge because I cut down rsi bloat significantly and had to update all of the map prototypes. I also **deleted centcomm, dart, and moonrise** because they are unmaintained and fixing their yml was much more CBT than the others (their gun containers already had magazines spawned, but some had duplicate magazines). Someone remind me to dump them in the archive repo. They need some sort of script run over them to clean them up.
- Guns all have a default spread for now until I make a debug overlay and we balance it. This really isn't worse than master atm.
- **There's probably going to be container oddities.** Detachable cells were removed from battery guns (per pjb in disc) so I deleted all of the containers but didn't get all of the children of those entities. These should easily show up in game and be deletable. I have already spent a lot of time fixing maps and I'm not spending more.

**Review:**
I'd recommend reviewing the following:
- GunSystem and the related components
- Check out a gun or two, e.g. SMGs + batteries
- Trust the yml linter to do its job on the rest

Huge credit to electro for getting predicted shooting working for me & stalker for doing the gun removals beforehand.

:cl: ElectroSr, metalgearsloth, & TStalkerZero
- add: Guns are now predicted. This should also lead to significantly fewer "Got late MsgEntity!" messages.
- add: Selector mode for guns has been added.
- remove: Many different guns have been removed to cull excess.
- remove: Revolver ammo counter has been removed making spinning much more exciting.
- remove: Battery weapons have had their ammo count and damage nerfed.
- remove: Gun bolts have been removed.
- tweak: Hitscan effects for guns are now fancy, using CEV-eris latest sprites.
- tweak: Gun accuracies have all been reset to a default and will be re-balanced in the future.
- tweak: Tasers are single shot until some sort of diminishing return on stuns is implemented, but should recharge faster.

### CODE REVIEW

**Comment by metalgearsloth** [comment]
Probably ready for review; ignore the Diff RSI action failure.

It's MVP so mappers can start working again and I'll do the rest.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            gun.NextFire += fireRate;
+            shots++;
+        }
+
+        // Get how many shots we're actually allowed to make, due to clip size or otherwise.
+        // Don't do this in the loop so we still reset NextFire.
+        switch (gun.SelectedMode)
+        {
+            case SelectiveFire.SemiAuto:
+                shots = Math.Min(shots, 1 - gun.ShotCounter);
```
Comment (wrexbe) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs:203]: Could this accidentally go negative, and trip that `DebugTools.Assert(shots >= 0);`  below? Not sure.

File: Resources/Textures/Objects/Weapons/Guns/Launchers/china_lake.rsi/meta.json [hunk 2]
```diff
@@ -10,15 +10,6 @@
     {
       "name": "icon"
     },
-    {
-      "name": "base"
```
Comment (wrexbe) [inline Resources/Textures/Objects/Weapons/Guns/Launchers/china_lake.rsi/meta.json:14]: The other ones are base, but this one has icon left? Is this on purpose?

File: Resources/Textures/Objects/Weapons/Guns/HMGs/minigun.rsi/meta.json [hunk 3]
```diff
@@ -7,45 +7,12 @@
     "y": 32
   },
   "states": [
-    {
-      "name": "base",
```
Comment (wrexbe) [inline Resources/Textures/Objects/Weapons/Guns/HMGs/minigun.rsi/meta.json:11]: Is this ok?

File: Resources/Textures/Objects/Weapons/Guns/Ammunition/Magazine/Pistol/calico_mag.rsi/meta.json [hunk 4]
```diff
@@ -9,12 +9,6 @@
     "states": [
         {
             "name": "icon"
```
Comment (wrexbe) [inline Resources/Textures/Objects/Weapons/Guns/Ammunition/Magazine/Pistol/calico_mag.rsi/meta.json:11]: ?

File: Resources/Prototypes/Objectives/traitorObjectives..yml [hunk 5]
```diff
@@ -169,5 +169,5 @@
         - DieCondition
```
Comment (wrexbe) [inline Resources/Prototypes/Objectives/traitorObjectives..yml:169]: This file is ..yml

**Review by wrexbe — CHANGES_REQUESTED** [review]

**Comment by wrexbe** [comment]
If I have harm on, and pick up a gun, I sometimes accidentally fire the weapon

File: Content.Client/Weapons/Ranged/Components/SpentAmmoVisualizer.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+[UsedImplicitly]
+public sealed class SpentAmmoVisualizer : AppearanceVisualizer
+{
+    [DataField("state")]
+    private string _state = "base";
+
+    public override void OnChangeData(AppearanceComponent component)
+    {
+        base.OnChangeData(component);
+        var sprite = IoCManager.Resolve<IEntityManager>().GetComponent<ISpriteComponent>(component.Owner);
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Client/Weapons/Ranged/Components/SpentAmmoVisualizer.cs:16]: It should be a visualizersystem instead if anything.

**Comment by metalgearsloth** [comment]
> If I have harm on, and pick up a gun, I sometimes accidentally fire the weapon

Feature :HECK:

File: Resources/Textures/Objects/Weapons/Guns/HMGs/minigun.rsi/meta.json (hunk 3 above)
Comment (metalgearsloth) [reply Resources/Textures/Objects/Weapons/Guns/HMGs/minigun.rsi/meta.json:11]: ye. Though I might add a visualizer for this so I will split out an icon.

File: Resources/Textures/Objects/Weapons/Guns/Launchers/china_lake.rsi/meta.json (hunk 2 above)
Comment (metalgearsloth) [reply Resources/Textures/Objects/Weapons/Guns/Launchers/china_lake.rsi/meta.json:14]: Yes, I generally did that for stuff with no mag states.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs:203]: It'll never hit here with fewer than 1 shot. Maybe if the selector mode is toggled while ShotCounter is active but in which case I'd want the assert tripping anyway.

File: Content.Shared/Containers/ItemSlot/ItemSlotsSystem.cs [hunk 7]
```diff
@@ -336,13 +340,13 @@ public bool TryEject(EntityUid uid, ItemSlot slot, EntityUid? user, [NotNullWhen
...
-            /// This handles logic with the slot itself
+            // This handles logic with the slot itself
             if (!CanEject(slot))
                 return false;
 
             item = slot.Item;
 
-            /// This handles user logic
+            // This handles user logic
             if (user != null && item != null && !_actionBlockerSystem.CanPickup(user.Value, item.Value))
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Shared/Containers/ItemSlot/ItemSlotsSystem.cs:350]: Not sure, not my code.

File: Content.Server/Content.Server.csproj [hunk 8]
```diff
@@ -25,6 +25,9 @@
     <ProjectReference Include="..\RobustToolbox\Robust.Server\Robust.Server.csproj" />
     <ProjectReference Include="..\Content.Shared\Content.Shared.csproj" />
   </ItemGroup>
+  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Content.Server.csproj:28]: Rider hours

File: Content.Shared/Weapons/Ranged/Events/AmmoShotEvent.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+namespace Content.Shared.Weapons.Ranged.Events;
+
+/// <summary>
+/// Raised when projectiles have been fired from a gun.
+/// </summary>
```
Comment (mirrorcult) [inline Content.Shared/Weapons/Ranged/Events/AmmoShotEvent.cs:5]: comments should specify what entities the event is raised on

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.ChamberMagazine.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+        if (TryComp<AppearanceComponent>(magEnt, out var magAppearance))
+        {
+            magAppearance.TryGetData<int>(AmmoVisuals.AmmoCount, out var addCount);
+            magAppearance.TryGetData<int>(AmmoVisuals.AmmoMax, out var addCapacity);
+            count += addCount;
+            capacity += addCapacity;
+        }
+
+        FinaliseMagazineTakeAmmo(uid, component, args, count, capacity, appearance);
```
Comment (mirrorcult) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.ChamberMagazine.cs:124]: ..why is there shared code that has actual logic dictated by stuff in appearance data? this seems weird

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Interactions.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        var cooldown = TimeSpan.FromSeconds(InteractNextFire);
+
+        if (component.NextFire < curTime)
+            component.NextFire = curTime + cooldown;
+        else
+            component.NextFire += cooldown;
+
+        PlaySound(component.Owner, component.SoundModeToggle?.GetSound(Random, ProtoManager), user);
+        Popup(Loc.GetString("gun-selected-mode", ("mode", GetLocSelector(fire))), component.Owner, user);
+        // When actions done add here.
```
Comment (mirrorcult) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Interactions.cs:71]: hm?

File: Content.Client/Weapons/Ranged/Components/MagVisualizer.cs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+using Content.Shared.Weapons.Ranged.Systems;
+using JetBrains.Annotations;
+using Robust.Client.GameObjects;
+using SharedGunSystem = Content.Shared.Weapons.Ranged.Systems.SharedGunSystem;
+
+namespace Content.Client.Weapons.Ranged.Components;
+
+[UsedImplicitly]
+public sealed class MagVisualizer : AppearanceVisualizer
+{
```
Comment (mirrorcult) [inline Content.Client/Weapons/Ranged/Components/MagVisualizer.cs:11]: this aint a visualizer system (unless this is just ported/moved and diff is lying)

File: Content.Client/Weapons/Ranged/Systems/GunSystem.AmmoCounter.cs [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+    /// <param name="component"></param>
+    private void RefreshControl(EntityUid uid, AmmoCounterComponent? component = null)
+    {
+        if (!Resolve(uid, ref component, false)) return;
+
+        component.Control?.Dispose();
+        component.Control = null;
+
+        var ev = new AmmoCounterControlEvent();
+        RaiseLocalEvent(uid, ev);
```
Comment (mirrorcult) [inline Content.Client/Weapons/Ranged/Systems/GunSystem.AmmoCounter.cs:36]:
```suggestion
        RaiseLocalEvent(uid, ev, false);
```

File: Content.Client/Weapons/Ranged/Systems/GunSystem.AmmoCounter.cs [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+    private void UpdateAmmoCount(EntityUid uid, AmmoCounterComponent component)
+    {
+        if (component.Control == null) return;
+
+        var ev = new UpdateAmmoCounterEvent()
+        {
+            Control = component.Control
+        };
+
+        RaiseLocalEvent(uid, ev);
```
Comment (mirrorcult) [inline Content.Client/Weapons/Ranged/Systems/GunSystem.AmmoCounter.cs:54]:
```suggestion
        RaiseLocalEvent(uid, ev, false);
```

File: Content.Client/Weapons/Ranged/Systems/GunSystem.ChamberMagazine.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    private void OnChamberMagazineAmmoUpdate(EntityUid uid, ChamberMagazineAmmoProviderComponent component, UpdateAmmoCounterEvent args)
+    {
+        if (args.Control is not ChamberMagazineStatusControl control) return;
+
+        var chambered = GetChamberEntity(uid);
+        var magEntity = GetMagazineEntity(uid);
+        var ammoCountEv = new GetAmmoCountEvent();
+
+        if (magEntity != null)
+            RaiseLocalEvent(magEntity.Value, ref ammoCountEv);
```
Comment (mirrorcult) [inline Content.Client/Weapons/Ranged/Systems/GunSystem.ChamberMagazine.cs:46]:
```suggestion
            RaiseLocalEvent(magEntity.Value, ref ammoCountEv, false);
```

File: Content.Client/Weapons/Ranged/Systems/GunSystem.Magazine.cs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            if (args.Control is DefaultStatusControl control)
+            {
+                control.Update(0, 0);
+            }
+
+            return;
+        }
+
+        RaiseLocalEvent(ent.Value, args);
```
[the remark this answers is not in the document]
Comment (mirrorcult) [reply Content.Client/Weapons/Ranged/Systems/GunSystem.Magazine.cs:27]: you get the idea

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Client/Weapons/Ranged/Components/MagVisualizer.cs (hunk 12 above)
Comment (metalgearsloth) [reply Content.Client/Weapons/Ranged/Components/MagVisualizer.cs:11]: It's just moved (the diff fell over).

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.ChamberMagazine.cs (hunk 10 above)
Comment (metalgearsloth) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.ChamberMagazine.cs:124]: lazy. there's an event for getting ammo count but I don't think I implemented it for every ammoprovider so I'll do that.

**Pull request merged**